From lethal@linux-sh.org Wed Feb 13 00:03:44 2008
From: Paul Mundt <lethal@linux-sh.org>
Date: Wed, 13 Feb 2008 17:02:33 +0900
Subject: USB: m66592-udc: reduce size of data structure.
To: Yoshihiro Shimoda <shimoda.yoshihiro@renesas.com>
Cc: Greg KH <greg@kroah.com>
Message-ID: <20080213080233.GA10847@linux-sh.org>
Content-Disposition: inline


Poking around with pahole, we see that m66592 handily shoves a u16 in
between larger types on 2 separate occasions leaving us with 2 2-byte
holes:

struct m66592 {
	...

	/* size: 1196, cachelines: 38 */
	/* sum members: 1192, holes: 2, sum holes: 4 */
	/* last cacheline: 12 bytes */
};	/* definitions: 1 */

Pairing them gets back 4-bytes:

struct m66592 {
	...

	/* size: 1192, cachelines: 38 */
	/* last cacheline: 8 bytes */
};	/* definitions: 1 */

Unfortunately it's not enough to save a cacheline with this massive
structure, but every byte helps.

Signed-off-by: Paul Mundt <lethal@linux-sh.org>
Signed-off-by: Yoshihiro Shimoda <shimoda.yoshihiro@renesas.com>
Signed-off-by: Greg Kroah-Hartman <gregkh@suse.de>

---
 drivers/usb/gadget/m66592-udc.h |    2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

--- a/drivers/usb/gadget/m66592-udc.h
+++ b/drivers/usb/gadget/m66592-udc.h
@@ -486,10 +486,10 @@ struct m66592 {
 
 	struct usb_request	*ep0_req;	/* for internal request */
 	u16			ep0_data;	/* for internal request */
+	u16			old_vbus;
 
 	struct timer_list	timer;
 
-	u16			old_vbus;
 	int			scount;
 
 	int			old_dvsq;
